flink-runtime/src/main/java/org/apache/flink/runtime/taskexecutor/TaskExecutor.java

```diff
@@ -2474,16 +2523,21 @@ public void jobManagerLostLeadership(final JobID jobId, final JobMasterId jobMas
                     "JobManager for job {} with leader id {} lost leadership.", jobId, jobMasterId);
 
             runAsync(
-                    () ->
-                            jobTable.getConnection(jobId)
-                                    .ifPresent(
-                                            jobManagerConnection ->
-                                                    disconnectJobManagerConnection(
-                                                            jobManagerConnection,
-                                                            new Exception(
-                                                                    "Job leader for job id "
-                                                                            + jobId
-                                                                            + " lost leadership."))));
+                    () -> {
+                        Optional<JobTable.Connection> connection = jobTable.getConnection(jobId);
+
+                        if (connection.isPresent()) {
+                            Exception cause =
+                                    new Exception(
+                                            "Job leader for job id " + jobId + " lost leadership.");
+                            if (isJobRecoveryEnabled()) {
+                                disconnectJobManagerConnectionAndCleanupPartitionLater(
```

Comment:
Would you confirm if it's possible to reconnect to the old JM? if it can happen, the retained partitions will be leaked.